### PR TITLE
Integrate Fastrace + Enable GZIP and ZSTD in Tonic

### PR DESCRIPTION
**Pull request by shehabgamin**
#169

### CODE REVIEW

File: crates/sail-telemetry/src/telemetry.rs [hunk 1]
```diff
@@ -1,86 +1,91 @@
...
+        let reporter = OpenTelemetryReporter::new(
+            opentelemetry_otlp::new_exporter()
+                .tonic()
+                .with_endpoint(url)
+                .with_protocol(opentelemetry_otlp::Protocol::Grpc)
+                .with_timeout(Duration::from_secs(
+                    opentelemetry_otlp::OTEL_EXPORTER_OTLP_TIMEOUT_DEFAULT,
+                ))
+                .build_span_exporter()
+                .expect("initialize oltp exporter"),
```
Comment (linhr) [inline crates/sail-telemetry/src/telemetry.rs]: Let's return error instead.

File: crates/sail-telemetry/src/telemetry.rs [hunk 2]
```diff
@@ -1,86 +1,91 @@
...
 }
 
-// TODO: init_metrics
-pub fn init_metrics() {}
+// We use DummyReporter instead of ConsoleReporter to avoid duplicate span output to the console.
+// Although using ConsoleReporter and suppressing span output in the logger is possible, it would
+// result in a different format.
+// Disabling fastrace when logging only to the console is another option, but retaining trace and
+// span IDs in the logs is useful.
+pub struct DummyReporter;
```
Comment (linhr) [inline crates/sail-telemetry/src/telemetry.rs:87]: Nice!

File: crates/sail-telemetry/src/telemetry.rs [hunk 3]
```diff
@@ -1,86 +1,91 @@
...
+impl Reporter for DummyReporter {
```
Comment (linhr) [inline crates/sail-telemetry/src/telemetry.rs:89]: Let's add one empty line before the `impl` block.

File: crates/sail-telemetry/src/trace_layer.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+use fastrace::prelude::*;
+use tonic::codegen::http::Request;
+use tower::{Layer, Service};
+
+#[derive(Clone)]
+pub struct TraceLayer {
+    name: &'static str, // TODO: Arc needed?
+}
+impl TraceLayer {
```
Comment (linhr) [inline crates/sail-telemetry/src/trace_layer.rs:14]: Let's add one empty line to separate top-level blocks in this file.

File: Cargo.toml [hunk 5]
```diff
@@ -69,6 +59,14 @@ aws-config = "1.5.5"
...
 clap = { version = "4.5.17", features = ["derive"] }
 num_enum = "0.7.3"
+log = "0.4.22"
+env_logger = "0.11.5"
+fastrace = { version = "0.7.2", features = ["enable"] }
+fastrace-opentelemetry = "0.7.2"
+opentelemetry = "0.24.0"
+opentelemetry_sdk = "0.24.1"
+opentelemetry-otlp = { version = "0.17.0", features = ["tls", "tls-roots"] }
+anyhow = "1.0.87"
```
Comment (linhr) [inline Cargo.toml]: Is this used anywhere?

**Review by linhr — APPROVED** [review]
This is great!

File: crates/sail-telemetry/src/telemetry.rs [hunk 6]
```diff
@@ -1,86 +1,91 @@
...
+pub fn init_logger(use_collector: bool) -> TelemetryResult<()> {
+    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("info"))
+        .format(move |buf, record| {
+            if use_collector {
+                Event::add_to_local_parent(record.level().as_str(), || {
+                    [("message".into(), record.args().to_string().into())]
+                });
+            }
+            let level = record.level();
+            let target = record.module_path().unwrap();
```
Comment (linhr) [inline crates/sail-telemetry/src/telemetry.rs]: Let's return error here as well.

File: Cargo.toml (hunk 5 above)
Comment (shehabgamin) [reply Cargo.toml]: nice catch, not anymore!

File: crates/sail-telemetry/src/telemetry.rs (hunk 6 above)
Comment (shehabgamin) [reply crates/sail-telemetry/src/telemetry.rs]: Nice catch! The `module_path` isn't guaranteed to be set so we should instead do `.unwrap_or("SOME DEFAULT");`

Comment (shehabgamin) [reply crates/sail-telemetry/src/telemetry.rs]: Nvm ignore this, we should be using `.target()` instead.

**Review by linhr — APPROVED** [review]

**Pull request merged**